helix-term/src/ui/mod.rs

```diff
@@ -145,15 +145,27 @@ pub fn file_picker(root: PathBuf, config: &helix_view::editor::Config) -> FilePi
     let files = walk_builder.build().filter_map(|entry| {
         let entry = entry.ok()?;
 
-        // This is faster than entry.path().is_dir() since it uses cached fs::Metadata fetched by ignore/walkdir
-        let is_dir = entry.file_type().map(|ft| ft.is_dir()).unwrap_or(false);
+        let file_type = entry.file_type();
+
+        // This is faster than entry.path().is_file() since it uses cached fs::Metadata fetched by ignore/walkdir
+        let is_file = file_type.as_ref().map_or(false, |ft| ft.is_file());
+
+        if is_file {
+            return Some(entry.into_path());
+        }
+
+        let is_dir = file_type.as_ref().map_or(false, |ft| ft.is_dir());
 
         if is_dir {
             // Will give a false positive if metadata cannot be read (eg. permission error)
             return None;
         }
 
-        Some(entry.into_path())
+        // Only list symlink if points to a file which we have permisson for
+        match std::fs::metadata(entry.path()).map(|meta| meta.is_file()) {
+            Err(_) | Ok(false) => None,
+            Ok(true) => Some(entry.into_path()),
+        }
```

Comment:
maybe just setting `follow_links(true)` on the builder in this case would be better, would fix #2246 as well...

Reply:
I think so, `fs::metadata` is way too costly to run per-entry (out of potentially thousands). I specifically removed code calling `fs::metadata` from the completion functions https://github.com/helix-editor/helix/commit/78fba8683bdce31ad2d8e710cb70fb9f05b21e8c#diff-a2f72be0dae9186774ac716ae568f5216a5223dbf91e6ac8a25f62617c4fa55cL144

Also, wouldn't this make more sense to do inside `editor.open`? Resolve the symlink, then either `set_error` "Not allowed to open the file" or open the file. I don't think files with insufficient permissions should just get hidden, it makes it seem like they're not there at all

Reply:
I've removed `fs::metadata` calls and changed to use `follow_symlinks()` instead. Added a configuration option to `FilePicker` to enable/disable following symlinks, so if it would lead to performance issues it can at least be toggled. Also error is now reported the same way like in `:cd` and `:open` if file picker fails. Hope I covers everything you mentioned.